src/metabase/api/advanced_computation/public.clj

```diff
@@ -6,14 +6,110 @@
+                                            (advcomp-common/run-query context query info))
+                                 out-chan (a/promise-chan (map api.public/transform-results))]
+                             (async.u/promise-pipe in-chan out-chan)
+                             out-chan)))
+           options)))
+
+(defn public-dashcard-results-async
+  "Return the results of running a query with `parameters` for Card with `card-id` belonging to Dashboard with
+  `dashboard-id`. Throws a 404 immediately if the Card isn't part of the Dashboard. Returns a `StreamingResponse`."
+  [dashboard-id card-id export-format parameters
+   & {:keys [context constraints]
+      :or   {context     :public-dashboard}}]
+  (api.public/check-card-is-in-dashboard card-id dashboard-id)
+  (let [params (api.public/resolve-params dashboard-id (if (string? parameters)
+                                                         (json/parse-string parameters keyword)
+                                                         parameters))]
+    (run-query-for-card-with-id-async
+     card-id export-format params
+     :dashboard-id dashboard-id
+     :context      context
+     :constraints  constraints)))
+
+(defn run-query-for-card-with-params-async
+  "Run the query associated with Card with `card-id` using JWT `token-params`, user-supplied URL `query-params`,
+   an `embedding-params` whitelist, and additional query `options`. Returns `StreamingResponse` that should be
+  returned as the API endpoint result."
+  {:style/indent 0}
+  [& {:keys [export-format card-id embedding-params token-params query-params options]}]
+  {:pre [(integer? card-id) (u/maybe? map? embedding-params) (map? token-params) (map? query-params)]}
+  (let [merged-id->value (api.embed/validate-and-merge-params embedding-params token-params (api.embed/normalize-query-params query-params))
+        parameters       (api.embed/apply-merged-id->value (api.embed/resolve-card-parameters card-id) merged-id->value)]
+    (apply run-query-for-card-with-id-async
+           card-id export-format parameters
+           :context :embedded-question, options)))
+
+(defn run-query-for-unsigned-token-async
+  "Run the query belonging to Card identified by `unsigned-token`. Checks that embedding is enabled both globally and
+  for this Card. Returns core.async channel to fetch the results."
+  [unsigned-token export-format query-params & options]
+  (let [card-id (eu/get-in-unsigned-token-or-throw unsigned-token [:resource :question])]
+    (api.embed/check-embedding-enabled-for-card card-id)
+    (run-query-for-card-with-params-async
+      :export-format     export-format
+      :card-id           card-id
+      :token-params      (eu/get-in-unsigned-token-or-throw unsigned-token [:params])
+      :embedding-params  (db/select-one-field :embedding_params Card :id card-id)
+      :query-params      query-params
+      :options           options)))
```

Comment:
The same advice applies here -- these functions are exactly the same as the versions they were copied from with the exception of a single line. They're already heavily parameterized so they can be used in a lot of different use cases. Just add a new parameter so you can override the `run-query-for-card-with-id-async` function that gets called rather then copying the entire implementation. 

It's better to have one function that can run embedded queries for 10 different use cases then to copypasta the same function 10 times for 10 different use cases